File: ctci/chapter_10_sorting_searching/python/10.2-group_anagrams.py
from typing import List


def sort_anagrams(arr: List[str]) -> List[str]:
    """
    Sorts an array of strings so that all anagrams are next to each other.
    """
    hash_table = {}
    for word in arr:
        sorted_word = "".join(sorted(word))
        hash_table.setdefault(sorted_word, []).append(word)
    # List comprehension to flatten list.
    return [v for s in hash_table.values() for v in s]


str_list = ["acre", "home", "race", "hack", "cram", "acer"]
print(sort_anagrams(str_list))
# ['acre', 'race', 'acer', 'home', 'hack', 'cram']

File: ctci/tmp/chapter_10_sorting_searching/js/10.2-groupAnagrams.js
/**
 * Sorts an array of strings so that all anagrams are next to each other.
 * @param   {string[]}  arr
 * @returns {string[]}
 */
const sort_anagrams = arr => {
  const map = {}
  for (let s of arr) {
    const sorted = s
      .split("")
      .sort()
      .join("")
    if (map[sorted] === undefined) map[sorted] = [s]
    else map[sorted].push(s)
  }
  return Object.values(map).reduce((a, b) => {
    a.push(...b)
    return a
  }, [])
}

const arr = ["acre", "home", "race", "hack", "cram", "acer"]
console.log(sort_anagrams(arr))
/* [ 'acre', 'race', 'acer', 'home', 'hack', 'cram' ] */
